File: SPV.CPP/SDK/Implement/MasterWallet.cpp
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
#include "IDChainSubWallet.h"
#include "EthSidechainSubWallet.h"
#include "MainchainSubWallet.h"
#include "SubWallet.h"
#include "MasterWallet.h"
#include "BTCSubWallet.h"
#include "RippleSubWallet.h"

#include <Plugin/Transaction/Asset.h>
#include <Common/Utils.h>
#include <Common/Log.h>
#include <Common/ErrorChecker.h>
#include <WalletCore/Mnemonic.h>
#include <WalletCore/CoinInfo.h>
#include <SpvService/Config.h>
#include <Wallet/WalletCommon.h>

#include <vector>
#include <boost/filesystem.hpp>
#include <ethereum/base/BREthereumLogic.h>
#include <ethereum/base/BREthereumAddress.h>

namespace fs = boost::filesystem;

namespace Elastos {
	namespace ElaWallet {

		MasterWallet::MasterWallet(const std::string &id,
								   const ConfigPtr &config,
								   const std::string &dataPath) :
				_id(id),
				_config(config) {

			_account = AccountPtr(new Account(dataPath + "/" + _id));
            SetupNetworkParameters();
		}

		MasterWallet::MasterWallet(const std::string &id,
								   const std::string &mnemonic,
								   const std::string &passphrase,
								   const std::string &passwd,
								   bool singleAddress,
								   const ConfigPtr &config,
								   const std::string &dataPath) :
				_id(id),
				_config(config) {

			_account = AccountPtr(new Account(dataPath + "/" + _id, mnemonic, passphrase, passwd, singleAddress));
			_account->Save();
            SetupNetworkParameters();
		}

        MasterWallet::MasterWallet(const std::string &id,
                                   const std::string &singlePrivateKey,
                                   const std::string &passwd,
                                   const ConfigPtr &config,
                                   const std::string &dataPath) :
                                   _id(id),
                                   _config(config) {
            _account = AccountPtr(new Account(dataPath + "/" + _id, singlePrivateKey, passwd));
            _account->Save();
            SetupNetworkParameters();
		}

		MasterWallet::MasterWallet(const std::string &id,
								   const nlohmann::json &keystoreContent,
								   const std::string &backupPassword,
								   const std::string &payPasswd,
								   const ConfigPtr &config,
								   const std::string &dataPath) :
				_id(id),
				_config(config) {

			KeyStore keystore;
			keystore.Import(keystoreContent, backupPassword);

			_account = AccountPtr(new Account(dataPath + "/" + _id, keystore, payPasswd));
			_account->Save();
            SetupNetworkParameters();
		}

//		MasterWallet::MasterWallet(const std::string &id,
//								   const nlohmann::json &readonlyWalletJson,
//								   const ConfigPtr &config,
//								   const std::string &dataPath) :
//			_id(id),
//			_config(config) {
//
//			_account = AccountPtr(new Account(dataPath + "/" + _id, readonlyWalletJson));
//			_account->Save();
//            SetupNetworkParameters();
//		}

		MasterWallet::MasterWallet(const std::string &id,
								   const std::vector<PublicKeyRing> &pubKeyRings,
								   uint32_t m,
								   const ConfigPtr &config,
								   const std::string &dataPath,
								   bool singleAddress,
								   bool compatible) :
				_id(id),
				_config(config) {
			ErrorChecker::CheckParam(pubKeyRings.size() < m, Error::InvalidArgument, "Invalid M");

			_account = AccountPtr(new Account(dataPath + "/" + _id, pubKeyRings, m, singleAddress, compatible));
			_account->Save();
            SetupNetworkParameters();
		}

		MasterWallet::MasterWallet(const std::string &id,
								   const std::string &xprv,
								   const std::string &payPassword,
								   const std::vector<PublicKeyRing> &cosigners,
								   uint32_t m,
								   const ConfigPtr &config,
								   const std::string &dataPath,
								   bool singleAddress,
								   bool compatible) :
				_id(id),
				_config(config) {

			ErrorChecker::CheckParam(cosigners.size() + 1 < m, Error::InvalidArgument, "Invalid M");

			_account = AccountPtr(new Account(dataPath + "/" + _id, xprv, payPassword, cosigners, m, singleAddress, compatible));
			_account->Save();
            SetupNetworkParameters();
		}

		MasterWallet::MasterWallet(const std::string &id,
								   const std::string &mnemonic,
								   const std::string &passphrase,
								   const std::string &payPasswd,
								   const std::vector<PublicKeyRing> &cosigners,
								   uint32_t m,
								   const ConfigPtr &config,
								   const std::string &dataPath,
								   bool singleAddress,
								   bool compatible) :
				_id(id),
				_config(config) {

			ErrorChecker::CheckParam(cosigners.size() + 1 < m, Error::InvalidArgument, "Invalid M");

			_account = AccountPtr(new Account(dataPath + "/" + _id, mnemonic, passphrase, payPasswd, cosigners, m, singleAddress, compatible));
			_account->Save();
            SetupNetworkParameters();
		}

		MasterWallet::~MasterWallet() {

		}

		std::string MasterWallet::GenerateMnemonic(const std::string &language, Mnemonic::WordCount wordCount) {
			return Mnemonic::Create(language, wordCount);
		}

		void MasterWallet::RemoveLocalStore() {
			_account->Remove();
		}

		std::string MasterWallet::GetID() const {
			return _id;
		}

		std::string MasterWallet::GetWalletID() const {
			return _id;
		}

		std::vector<ISubWallet *> MasterWallet::GetAllSubWallets() const {
			ArgInfo("{} {}", _id, GetFunName());

			std::vector<ISubWallet *> subwallets;
			for (WalletMap::const_iterator it = _createdWallets.cbegin(); it != _createdWallets.cend(); ++it) {
				subwallets.push_back(it->second);
			}

			std::string result;
			for (size_t i = 0; i < subwallets.size(); ++i)
				result += subwallets[i]->GetChainID() + ",";

			ArgInfo("r => {}", result);
			return subwallets;
		}

		ISubWallet *MasterWallet::GetSubWallet(const std::string &chainID) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("chainID: {}", chainID);

			if (_createdWallets.find(chainID) != _createdWallets.end()) {
				return _createdWallets[chainID];
			}

			return nullptr;
		}

		ISubWallet *MasterWallet::CreateSubWallet(const std::string &chainID) {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("chainID: {}", chainID);

			ErrorChecker::CheckParamNotEmpty(chainID, "Chain ID");
			ErrorChecker::CheckParam(chainID.size() > 128, Error::InvalidArgument, "Chain ID sould less than 128");

			if (_createdWallets.find(chainID) != _createdWallets.end()) {
				ISubWallet *subWallet = _createdWallets[chainID];
				ArgInfo("r => already created");
				return subWallet;
			}

			ChainConfigPtr chainConfig = _config->GetChainConfig(chainID);
			ErrorChecker::CheckLogic(chainConfig == nullptr, Error::InvalidArgument, "Unsupport chain ID: " + chainID);

			CoinInfoPtr info(new CoinInfo());

			info->SetChainID(chainID);

			ISubWallet *subWallet = SubWalletFactoryMethod(info, chainConfig, this, _config->GetNetType());
			_createdWallets[chainID] = subWallet;
			_account->AddSubWalletInfoList(info);
			_account->Save();

			return subWallet;
		}

		bool MasterWallet::VerifyPrivateKey(const std::string &mnemonic, const std::string &passphrase) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("mnemonic: *");
			ArgInfo("passphrase: *");
			bool r = _account->VerifyPrivateKey(mnemonic, passphrase);
			ArgInfo("r => {}", r);
			return r;
		}

		bool MasterWallet::VerifyPassPhrase(const std::string &passphrase, const std::string &payPasswd) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("passphrase: *");
			ArgInfo("payPasswd: *");
			bool r = _account->VerifyPassPhrase(passphrase, payPasswd);
			ArgInfo("r => {}", r);
			return r;
		}

		bool MasterWallet::VerifyPayPassword(const std::string &payPasswd) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("payPasswd: *");
			bool r = _account->VerifyPayPassword(payPasswd);
			ArgInfo("r => {}", r);
			return r;
		}

		void MasterWallet::CloseAllSubWallets() {
			for (WalletMap::iterator it = _createdWallets.begin(); it != _createdWallets.end(); ) {
				ISubWallet *subWallet = it->second;
				std::string id = _id + ":" + subWallet->GetChainID();
				Log::info("{} closing...", id);

				it = _createdWallets.erase(it);

				delete subWallet;
				subWallet = nullptr;
				Log::info("{} closed", id);
			}
		}

        void MasterWallet::SetupNetworkParameters() {
            const std::map<std::string, ChainConfigPtr> configs = _config->GetConfigs();

            for (std::map<std::string, ChainConfigPtr>::const_iterator it = configs.begin(); it != configs.end(); ++it)
                if (!it->second->Name().empty())
                    InsertEthereumNetwork(it->second->Name().c_str(), it->second->ChainID(), it->second->NetworkID());
        }

		void MasterWallet::DestroyWallet(const std::string &chainID) {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("chainID: {}", chainID);

			if (_createdWallets.find(chainID) == _createdWallets.end())
				ErrorChecker::ThrowParamException(Error::InvalidArgument, "chainID not found");

			ISubWallet *subWallet = _createdWallets[chainID];
			_account->RemoveSubWalletInfo(subWallet->GetChainID());
			_account->Save();

			_createdWallets.erase(chainID);

			delete subWallet;
			subWallet = nullptr;

			ArgInfo("r => {} {} done", _id, GetFunName());
		}

		nlohmann::json MasterWallet::GetPubKeyInfo() const {
			ArgInfo("{} {}", _id, GetFunName());

			nlohmann::json j = _account->GetPubKeyInfo();

			ArgInfo("r => {}", j.dump());
			return j;
		}

#if 0
		nlohmann::json MasterWallet::ExportReadonlyWallet() const {
			ArgInfo("{} {}", _id, GetFunName());

			nlohmann::json j = _account->ExportReadonlyWallet();

			ArgInfo("r => {}", j.dump());
			return j;
		}
#endif

		nlohmann::json MasterWallet::ExportKeystore(const std::string &backupPassword, const std::string &payPassword) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("backupPassword: *");
			ArgInfo("payPassword: *");
			ErrorChecker::CheckPassword(backupPassword, "Backup");

			std::vector<CoinInfoPtr> coinInfo = _account->SubWalletInfoList();
			_account->SetSubWalletInfoList(coinInfo);
			_account->Save();

			KeyStore keyStore = _account->ExportKeystore(payPassword);
			nlohmann::json j = keyStore.Export(backupPassword, true);

			ArgInfo("r => *");
			return j;
		}

		std::string MasterWallet::ExportMnemonic(const std::string &payPassword) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("payPassword: *");

			std::string mnemonic = _account->ExportMnemonic(payPassword);

			ArgInfo("r => *");
			return mnemonic;
		}

		std::string MasterWallet::ExportPrivateKey(const std::string &payPasswd) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("payPsswd: *");

			ErrorChecker::CheckLogic(_account->Readonly(), Error::UnsupportOperation,
									 "Unsupport operation: read-only wallet do not contain xprv");

			std::string xprv = _account->GetxPrvKeyString(payPasswd);

			ArgInfo("r => *");
			return xprv;
		}

		std::string MasterWallet::ExportMasterPublicKey() const {
			ArgInfo("{} {}", _id, GetFunName());

			std::string mpk = _account->MasterPubKeyString();

			ArgInfo("r => {}", mpk);
			return mpk;
		}

        void MasterWallet::InitSubWallets() {
            const std::vector<CoinInfoPtr> &info = _account->SubWalletInfoList();

            for (int i = 0; i < info.size(); ++i) {
                ChainConfigPtr chainConfig = _config->GetChainConfig(info[i]->GetChainID());
                if (chainConfig == nullptr) {
                    Log::error("Can not find config of chain ID: " + info[i]->GetChainID());
                    continue;
                }

                ISubWallet *subWallet = SubWalletFactoryMethod(info[i], chainConfig, this, _config->GetNetType());
                ErrorChecker::CheckCondition(subWallet == nullptr, Error::CreateSubWalletError,
                                             "Recover sub wallet error");
                _createdWallets[subWallet->GetChainID()] = subWallet;
            }
        }

		ISubWallet *MasterWallet::SubWalletFactoryMethod(const CoinInfoPtr &info, const ChainConfigPtr &config,
														MasterWallet *parent, const std::string &netType) {

			if (info->GetChainID() == "ELA") {
				return new MainchainSubWallet(info, config, parent, netType);
			} else if (info->GetChainID() == "IDChain") {
                return new IDChainSubWallet(info, config, parent, netType);
            } else if (info->GetChainID() == "BTC") {
			    return new BTCSubWallet(info, config, parent, netType);
			} else if (info->GetChainID().find("ETH") !=  std::string::npos) {
                return new EthSidechainSubWallet(info, config, parent, netType);
//            } else if (info->GetChainID() == "XRP") {
//			    return new RippleSubWallet(info, config, parent, netType);
			} else {
				ErrorChecker::ThrowLogicException(Error::InvalidChainID, "Invalid chain ID: " + info->GetChainID());
			}

			return nullptr;
		}

		std::string MasterWallet::GetDataPath() const {
			return _account->GetDataPath();
		}

		AccountPtr MasterWallet::GetAccount() const {
			return _account;
		}

		bool MasterWallet::IsAddressValid(const std::string &address) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("addr: {}", address);

			bool valid = Address(address).Valid();
			if (!valid) {
				valid = addressValidateString(address.c_str()) == ETHEREUM_BOOLEAN_TRUE;
			}

			ArgInfo("r => {}", valid);
			return valid;
		}

		bool MasterWallet::IsSubWalletAddressValid(const std::string &chainID, const std::string &address) const {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("chainID: {}", chainID);
			ArgInfo("address: {}", address);

            bool valid = false;
            if (chainID == CHAINID_MAINCHAIN || chainID == CHAINID_IDCHAIN || chainID == CHAINID_TOKENCHAIN) {
                valid = Address(address).Valid();
            } else if (chainID == "BTC") {
                BRAddressParams addrParams;
                if (_config->GetNetType() == CONFIG_MAINNET) {
                    addrParams = BITCOIN_ADDRESS_PARAMS;
                } else {
                    addrParams = BITCOIN_TEST_ADDRESS_PARAMS;
                }
                valid = BRAddressIsValid(addrParams, address.c_str());
            } else if (chainID.find("ETH") != std::string::npos) {
                valid = addressValidateString(address.c_str()) == ETHEREUM_BOOLEAN_TRUE;
            }

			ArgInfo("r => {}", valid);
			return valid;
		}

		std::vector<std::string> MasterWallet::GetSupportedChains() const {
			ArgInfo("{} {}", _id, GetFunName());

			std::vector<std::string> chainIDs = _config->GetAllChainIDs();

			std::string result;
			for (size_t i = 0; i < chainIDs.size(); ++i) {
				result += chainIDs[i] + ", ";
			}

			ArgInfo("r => {}", result);
			return chainIDs;
		}

		void MasterWallet::ChangePassword(const std::string &oldPassword, const std::string &newPassword) {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("old: *");
			ArgInfo("new: *");

			_account->ChangePassword(oldPassword, newPassword);
		}

		void MasterWallet::ResetPassword(const std::string &mnemonic, const std::string &passphrase,
										 const std::string &newPassword) {
			ArgInfo("{} {}", _id, GetFunName());
			ArgInfo("m: *");
			ArgInfo("passphrase: *");
			ArgInfo("passwd: *");

			_account->ResetPassword(mnemonic, passphrase, newPassword);

			ArgInfo("r => ");
		}

		nlohmann::json MasterWallet::GetBasicInfo() const {
			ArgInfo("{} {}", _id, GetFunName());

			nlohmann::json info = _account->GetBasicInfo();

			ArgInfo("r => {}", info.dump());
			return info;
		}

		bool MasterWallet::IsEqual(const MasterWallet &wallet) const {
			return _account->Equal(wallet._account);
		}

		void MasterWallet::FlushData() {
			for (WalletMap::const_iterator it = _createdWallets.cbegin(); it != _createdWallets.cend(); ++it) {
				SubWallet *subWallet = dynamic_cast<SubWallet*>(it->second);
				if (subWallet)
					subWallet->FlushData();
			}
		}

		ChainConfigPtr MasterWallet::GetChainConfig(const std::string &chainID) const {
			return _config->GetChainConfig(chainID);
		}

	}
}

File: UnionSquare.Android/app/src/main/java/org/elastos/wallet/core/MasterWallet.java
// Copyright (c) 2012-2019 The Elastos Open Source Project
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

package org.elastos.wallet.core;

import org.elastos.wallet.ela.utils.Log;

import java.util.ArrayList;

/**
 * MasterWallet
 */
public class MasterWallet {
    static public class CHAINID {
        public static String MAIN = "ELA";
        public static String ID = "IDChain";
    }

    static public String TAG = "IMasterWallet";

    private long mInstance;
    private ArrayList<SubWallet> mSubWallets = new ArrayList<SubWallet>();

    public MasterWallet(long instance) throws WalletException {
        mInstance = instance;
        Object[] allSubWallets = GetAllSubWallets(mInstance);
        for (int i = 0; i < allSubWallets.length; i++) {
            mSubWallets.add((SubWallet) allSubWallets[i]);
        }
    }

    public String GetID() {
        return GetID(mInstance);
    }

    public String GetBasicInfo() throws WalletException {
        return GetBasicInfo(mInstance);
    }

    public ArrayList<SubWallet> GetAllSubWallets() throws WalletException {
        Object[] allSubWallets = GetAllSubWallets(mInstance);

        for (int i = 0; i < allSubWallets.length; i++) {
            SubWallet subWallet = (SubWallet) allSubWallets[i];
            boolean found = false;
            for (int j = 0; j < mSubWallets.size(); ++j) {
                if (mSubWallets.get(j).GetChainID().equals(subWallet.GetChainID())) {
                    found = true;
                    break;
                }
            }
            if (!found)
                mSubWallets.add(subWallet);
        }

        return mSubWallets;
    }

    public SubWallet GetSubWallet(String chainID) throws WalletException {
        for (int i = 0; i < mSubWallets.size(); i++) {
            if (chainID.equals(mSubWallets.get(i).GetChainID())) {
                return mSubWallets.get(i);
            }
        }

        return null;
    }

    public SubWallet CreateSubWallet(String chainID) throws WalletException {
        if ((!CHAINID.MAIN.equals(chainID)) && (!CHAINID.ID.equals(chainID))) {
            throw new WalletException("Not support the other sidechain now.");
        }

        for (int i = 0; i < mSubWallets.size(); ++i) {
            if (mSubWallets.get(i).GetChainID().equals(chainID)) {
                return mSubWallets.get(i);
            }
        }

        long subProxy = CreateSubWallet(mInstance, chainID);
        if (subProxy == 0) {
            Log.e(TAG, "Native create subwallet fail: subProxy is 0");
            throw new WalletException("Native create subwallet fail");
        }

        SubWallet subWallet = null;

        if (CHAINID.MAIN.equals(chainID)) {
            subWallet = new MainchainSubWallet(subProxy);
        } else if (CHAINID.ID.equals(chainID)) {
            subWallet = new IDChainSubWallet(subProxy);
        } else {
            throw new WalletException("Unsupport chainID: " + chainID);
        }

        mSubWallets.add(subWallet);

        return subWallet;
    }

    public void DestroyWallet(SubWallet wallet) throws WalletException {
        for (int i = 0; i < mSubWallets.size(); ++i) {
            if (mSubWallets.get(i).GetChainID().equals(wallet.GetChainID())) {
                mSubWallets.remove(i);
                break;
            }
        }
        wallet.RemoveCallback();
        DestroyWallet(mInstance, wallet.GetChainID());
    }

    public String GetOwnerPublicKeyRing() throws WalletException {
        return GetOwnerPublicKeyRing(mInstance);
    }

    public String GetPublicKeyRing() throws WalletException {
        return GetPublicKeyRing(mInstance);
    }

    public boolean IsAddressValid(String address) throws WalletException {
        return IsAddressValid(mInstance, address);
    }

    public String[] GetSupportedChains() throws WalletException {
        return GetSupportedChains(mInstance);
    }

    public long GetInstance() {
        return mInstance;
    }

    public void ChangePassword(String oldPassword, String newPassword) throws WalletException {
        ChangePassword(mInstance, oldPassword, newPassword);
    }

    public String GetPubKeyInfo() throws WalletException {
        return GetPubKeyInfo(mInstance);
    }

    public boolean VerifyPrivateKey(String mnemonic, String passphrase) throws WalletException {
        return VerifyPrivateKey(mInstance, mnemonic, passphrase);
    }

    public boolean VerifyPassPhrase(String passPhrase, String payPasswd) throws WalletException {
        return VerifyPassPhrase(mInstance, passPhrase, payPasswd);
    }

    public boolean VerifyPayPassword(String payPasswd) throws WalletException {
        return VerifyPayPassword(mInstance, payPasswd);
    }

    public String ExportKeystore(String backPasswd, String payPasswd) throws WalletException {
        return ExportKeystore(mInstance, backPasswd, payPasswd);
    }

    public String ExportMnemonic(String payPasswd) throws WalletException {
        return ExportMnemonic(mInstance, payPasswd);
    }

    public String ExportReadonlyWallet() throws WalletException {
        return ExportReadonlyWallet(mInstance);
    }

    public String ExportPrivateKey(String payPasswd) throws WalletException {
        return ExportPrivateKey(mInstance, payPasswd);
    }

    public String ExportMasterPublicKey() throws WalletException {
        return ExportMasterPublicKey(mInstance);
    }

    private native String GetID(long instance);

    private native String GetBasicInfo(long instance);

    private native Object[] GetAllSubWallets(long instance);

    private native long GetSubWallet(long instance, String chainID);

    private native long CreateSubWallet(long instance, String chainID);

    private native String GetOwnerPublicKeyRing(long instance);

    private native String GetPublicKeyRing(long instance);

    private native void DestroyWallet(long instance, String chainID);

    private native boolean IsAddressValid(long instance, String address);

    private native String[] GetSupportedChains(long instance);

    private native void ChangePassword(long instance, String oldPassword, String newPassword);

    private native String GetPubKeyInfo(long instance);

    private native boolean VerifyPrivateKey(long instance, String mnemonic, String passphrase);

    private native boolean VerifyPassPhrase(long instance, String passPhrase, String payPasswd);

    private native boolean VerifyPayPassword(long instance, String payPasswd);

    private native String ExportKeystore(long instance, String backPasswd, String payPasswd);

    private native String ExportMnemonic(long instance, String payPasswd);

    private native String ExportReadonlyWallet(long instance);

    private native String ExportPrivateKey(long instance, String payPasswd);

    private native String ExportMasterPublicKey(long instance);
}
